File: packages/hquant/src/indicator/atr.ts
import { CircularQueue } from "../common/CircularQueue";
import { TypedRingBuffer } from "../common/TypedRingBuffer";
import { Kline, Indicator } from "../interface";

/**
 * 平均真实波幅（ATR）指标
 */
export class ATR implements Indicator {
  buffer: CircularQueue<Kline>;
  period: number;
  result: TypedRingBuffer;
  maxHistoryLength = 120;

  constructor({ period, maxHistoryLength }: { period: number; maxHistoryLength?: number }) {
    this.maxHistoryLength = maxHistoryLength || this.maxHistoryLength;
    this.period = period;
    this.buffer = new CircularQueue<Kline>(period);
    this.result = new TypedRingBuffer('float', this.maxHistoryLength);
  }

  // 计算单个K线的真实波幅
  private getTrueRange(curr: Kline, prev?: Kline): number {
    if (!prev) return curr.high - curr.low;
    return Math.max(
      curr.high - curr.low,
      Math.abs(curr.high - prev.close),
      Math.abs(curr.low - prev.close)
    );
  }
  calc(): number {
    let trSum = 0;
    for (let i = 0; i < this.period; i++) {
      const curr = this.buffer.get(i);
      const prev = i > 0 ? this.buffer.get(i - 1) : undefined;
      trSum += this.getTrueRange(curr, prev);
    }
    const atr = trSum / this.period;
    return atr;
  }
  add(data: Kline) {
    this.buffer.push(data);
    if (this.buffer.size() < this.period) {
      this.result.push(NaN);
      return;
    }
    this.result.push(this.calc());
  }

  updateLast(data: Kline) {
    if (this.buffer.size() === 0) return;
    this.buffer.update(this.buffer.size() - 1, data);
    this.result.update(this.result.size() - 1, this.calc());
  }

  getValue(index: number = -1): number {
    return this.result.get(index < 0 ? this.result.size() - 1 : index);
  }
}

File: packages/hquant-go-v2/indicator/atr.go

package indicator

import (
	"math"

	"hquant-go-v2/common"
"hquant-go-v2/types"
)

// ATR represents the Average True Range indicator.
type ATR struct {
	buffer *common.CircularQueue
	result *common.TypedRingBuffer
	period int
}

// NewATR creates a new ATR indicator.
func NewATR(period, maxHistoryLength int) *ATR {
	return &ATR{
		buffer: common.NewCircularQueue(period),
		result: common.NewTypedRingBuffer(maxHistoryLength),
		period: period,
	}
}

// getTrueRange calculates the true range for a given Kline.
func (a *ATR) getTrueRange(curr hquant.Kline, prev *hquant.Kline) float64 {
	if prev == nil {
		return curr.High - curr.Low
	}
	return math.Max(
		curr.High-curr.Low,
		math.Abs(curr.High-prev.Close),
		math.Abs(curr.Low-prev.Close),
	)
}

// calc calculates the ATR value.
func (a *ATR) calc() float64 {
	trSum := 0.0
	for i := 0; i < a.buffer.Size(); i++ {
		curr := a.buffer.Get(i).(hquant.Kline)
		var prev *hquant.Kline
		if i > 0 {
			p := a.buffer.Get(i - 1).(hquant.Kline)
			prev = &p
		}
		trSum += a.getTrueRange(curr, prev)
	}
	return trSum / float64(a.period)
}

// Add adds a new data point to the indicator.
func (a *ATR) Add(data hquant.Kline) {
	a.buffer.Push(data)
	if a.buffer.Size() < a.period {
		a.result.Push(math.NaN())
		return
	}
	a.result.Push(a.calc())
}

// UpdateLast updates the last data point of the indicator.
func (a *ATR) UpdateLast(data hquant.Kline) {
	if a.buffer.Size() == 0 {
		return
	}
	if a.buffer.Size() < a.period {
		return
	}
	a.buffer.Update(a.buffer.Size()-1, data)
	a.result.Update(a.result.Size()-1, a.calc())
}

// GetValue returns the ATR value at a specific index.
func (a *ATR) GetValue(index int) interface{} {
	return a.result.Get(index)
}
